torch/csrc/dynamo/eval_frame.c

```diff
@@ -798,19 +748,84 @@ inline static PyObject* eval_custom_code_impl(
+  //  3. CPython interpreter executes the code object
+  //
+  // Dynamo hooks the 3th step: before executing the code object, Dynamo transforms the code object into a new code object. Then, the old frame is not suitable for executing the new code. Therefore, Dynamo needs to manually create and initialize a new frame to execute the new code.
+  // The main task is to copy data in old frame to new frame, concerning a storage space named `localsplus`.
+  //
+  // localsplus storage is an array with the following layout:
+  // |   args   |   new_locals    |    cell_variables |   free_variables    |
+  // | <--- from left to right, index from 0 to n - 1 ---> |
+  // code.co_varnames == args + new_locals, code.co_nlocals == len(code.co_varnames)
+  // code.co_freevars == free_variables
+  // In Python 3.10 and lower, `n == code.co_nlocals + len(code.co_cellvars) + len(code.co_freevars)` (Python expression)
+  // In Python 3.11 and higher, `n <= code.co_nlocals + len(code.co_cellvars) + len(code.co_freevars)` (Python expression). There is an extra field in Python C-API: `n == code->co_nlocalsplus` (C expression) to retrieve the length of array.
+  // The complexity happens if an argument becomes a cell variable:
+  //  In Python 3.10 and lower, `code.co_cellvars == cell_variables`, and the corresponding slot in args becomes `NULL`.
+  //  In Python 3.11 and higher, `code.co_cellvars > cell_variables`, that cell variable is still stored in args, with a flag set in corresponding item's `co_localspluskinds` .
+  //
+  // ideally, we need to look up new localsplus from old localsplus by name:
+  // for i, name, value in enumerate(localsplusnames_old):
+  //   if value != NULL: (NULL happens for new local variables and arguments that becomes cell variables)
+  //     name_to_idx[name] = i
+  // for i, name in enumerate(localsplusnames_new):
+  //  if name in name_to_idx:
+  //    fastlocals_new[i] = fastlocals_old[name_to_idx[name]]
+  //
+  // The above process of building a `name_to_idx` mapping is expensive.
+  // Dynamo makes the following assumptions:
+  //  1. new code has the same arguments as the old code (both the number and the order)
+  //  2. new code has the same cell variables as the old code (both the number and the order)
+  //  3. new code has the same free variables as the old code (both the number and the order)
+  //  The only flexibility lies in new local variables: new code can introduce their own variables.
+  // With these assumptions, Dynamo can copy data directly by index. Dynamo just needs to take care of copying cell variables correctly.
+  // To avoid runtime cost, the assumptions are checked when we first generate the code object in pytorch/torch/_dynamo/convert_frame.py .
+
+
+  // copy args
+  Py_ssize_t total_argcount_old = frame->f_code->co_argcount + frame->f_code->co_kwonlyargcount + !!(frame->f_code->co_flags & CO_VARARGS) + !!(frame->f_code->co_flags & CO_VARKEYWORDS);
```

Comment:
Can you write a comment that varargs/varkwargs need to be handled separately? Also, can we expand the `!!` notation? It can be confusing to read at a glance. Also consider adding a test for bytecode with varargs and varkwargs.

Reply:
Fixed in [5db055e](https://github.com/pytorch/pytorch/pull/115062/commits/5db055e271f4c1f1d8a3d219d25c142af3ba1788).